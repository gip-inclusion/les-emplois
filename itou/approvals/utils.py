from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
