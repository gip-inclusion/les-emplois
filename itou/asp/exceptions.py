class ASPReferenceError(Exception):
    "Generic error class for what may go wrong in this app (a lot of things actually)"


class UnknownCommuneError(ASPReferenceError):
    "Unable to lookup commune. May happen from time to time as itou and ASP INSEE references are not synced"


class CommuneUnknownInPeriodError(ASPReferenceError):
    "Unable to find a commune for a given period"
