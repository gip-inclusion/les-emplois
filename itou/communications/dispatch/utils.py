class JobSeekerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_job_seeker


class EmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_employer


class PrescriberNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_prescriber


class PrescriberOrEmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and (self.user.is_prescriber or self.user.is_employer)
