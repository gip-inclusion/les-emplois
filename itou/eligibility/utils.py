from collections import Counter


def geiq_allowance_amount(is_authorized_prescriber, administrative_criteria) -> int:
    """Amount of granted allowance for job seeker.

    Calculated in function of:
        - author kind
        - number, annex and level of administrative criteria.
    Currently, only 3 amounts possible:
        - 0
        - 814EUR
        - 1400EUR
    """

    # Even when no criteria in the diagnosis
    if is_authorized_prescriber:
        return 1400

    # Count by annex
    annex_cnt = Counter(c.annex for c in administrative_criteria)

    # Only annex 2 administrative criteria have a level defined
    level_cnt = Counter(c.level for c in administrative_criteria if c.level)

    # At least one level 1 criterion or at least two level 2 criteria
    if level_cnt["1"] > 0 or level_cnt["2"] > 1:
        return 1400

    # Criteria in both annex ("1+2") must be counted as annex 1
    if annex_cnt["1"] + annex_cnt["1+2"] > 0:
        return 814

    return 0
