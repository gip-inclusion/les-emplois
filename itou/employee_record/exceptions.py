class SerializationError(Exception):
    """Mainly raised during serialization phases in employee record management commands."""


class InvalidStatusError(Exception):
    """Raised when changing status from an invalid previous status value."""
