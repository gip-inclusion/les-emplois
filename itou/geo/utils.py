from typing import Optional

from django.contrib.gis.geos import GEOSGeometry


def coords_to_geometry(lat, lon) -> Optional[GEOSGeometry]:
    if lat is not None and lon is not None:
        return GEOSGeometry(f"POINT({lon} {lat})")
    return None


def multipolygon_to_geometry(spec: str) -> GEOSGeometry:
    return GEOSGeometry(f"{spec}")
