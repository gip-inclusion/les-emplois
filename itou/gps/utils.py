import logging
from typing import NamedTuple

import pandas as pd
from django.core.cache import cache

from itou.gps.models import FranceTravailContact
from itou.utils.python import timeit


logger = logging.getLogger(__name__)


class FranceTravailContactDetails(NamedTuple):
    prescriber_name: str
    prescriber_email: str


GPS_ADVISORS_KEY = "GPS_ADVISORS"


@timeit
def parse_gps_advisors_file(import_file):
    # NOM | PRENOM | NIR | DATE DE NAISSANCE | ASSEDIC | DC_STRUCTUREPRINCIPALEDE | DC_AGENTREFERENT
    # DC_STRUCTURERATTACH | DC_NOMAGENTREFERENT | DC_MAIL | DC_LBLPOSITIONNEMENTIAE
    df = pd.read_excel(
        import_file,
        converters={
            "NIR": str,
            "DC_NOMAGENTREFERENT": str,
            "DC_MAIL": str,
        },
    )

    df = df.rename(columns={"NIR": "nir", "DC_NOMAGENTREFERENT": "prescriber_name", "DC_MAIL": "prescriber_email"})
    df = df[["nir", "prescriber_name", "prescriber_email"]]
    df = df.dropna(subset=["prescriber_name", "prescriber_email"])
    logger.info(f"Found {len(df)} rows from GPS export.")

    nir_to_contact = {}
    count_nir_invalid_length = 0
    for row in df.itertuples():
        nir = row.nir
        match len(nir):
            case 13:
                # not all of the NIRs in the dataset are complete, so we treat them
                nir = f"{nir}{str(97 - int(nir) % 97).zfill(2)}"
            case 15:
                pass
            case _:
                count_nir_invalid_length += 1
                continue
        contact_details = FranceTravailContactDetails(row.prescriber_name, row.prescriber_email)
        existing = nir_to_contact.setdefault(nir, contact_details)
        if existing != contact_details:
            logger.warning(
                f"JobSeekerProfile with nir {nir} matched an additional contact that have not been imported."
            )

    if count_nir_invalid_length:
        logger.warning(f"There are {count_nir_invalid_length} included NIR values of invalid length after treatment.")

    cache.set(GPS_ADVISORS_KEY, nir_to_contact)
    return nir_to_contact


def create_or_update_advisor(jobseeker_profile, nir_to_contact, commit=True):
    if jobseeker_profile.nir not in nir_to_contact:
        return None, False

    contact_name, contact_email = nir_to_contact[jobseeker_profile.nir]
    created = False
    # prepare to create or update FranceTravailContact
    try:
        advisor = jobseeker_profile.advisor_information
        advisor.name = contact_name
        advisor.email = contact_email
    except FranceTravailContact.DoesNotExist:
        advisor = FranceTravailContact(
            name=contact_name,
            email=contact_email,
            jobseeker_profile=jobseeker_profile,
        )
        created = True
    if commit:
        advisor.save()

    return advisor, created


def find_job_seeker_advisor(jobseeker_profile):
    if not jobseeker_profile.nir:
        return

    nir_to_contact = cache.get(GPS_ADVISORS_KEY)

    # No cache available : wait for the next cron call (in less than 15 minutes)
    if nir_to_contact is None:
        return

    create_or_update_advisor(jobseeker_profile, nir_to_contact)
