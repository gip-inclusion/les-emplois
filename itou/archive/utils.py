import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
