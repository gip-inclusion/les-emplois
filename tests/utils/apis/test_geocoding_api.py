import pytest
from django.contrib.gis.geos import GEOSGeometry

from itou.utils.apis import geocoding
from itou.utils.apis.exceptions import GeocodingDataError
from itou.utils.mocks.geocoding import BAN_GEOCODING_API_NO_RESULT_MOCK, BAN_GEOCODING_API_WITH_RESULT_RESPONSE


def test_get_geocoding_data(caplog, snapshot, respx_mock, settings):
    settings.API_BAN_BASE_URL = "https://geo.foo"
    respx_mock.get(f"{settings.API_BAN_BASE_URL}/search/").respond(200, json=BAN_GEOCODING_API_WITH_RESULT_RESPONSE)

    result = geocoding.get_geocoding_data("")
    # Expected data comes from BAN_GEOCODING_API_RESULT_MOCK.
    assert result == {
        "score": 0.5197687103594081,
        "address_line_1": "10 Place des Cinq Martyrs du Lycée Buffon",
        "number": "10",
        "lane": "Place des Cinq Martyrs du Lycée Buffon",
        "address": "10 Place des Cinq Martyrs du Lycée Buffon",
        "post_code": "75015",
        "insee_code": "75115",
        "city": "Paris",
        "longitude": 2.316754,
        "latitude": 48.838411,
        "coords": GEOSGeometry("POINT(2.316754 48.838411)"),
    }
    assert [record for record in caplog.record_tuples if record[0] == geocoding.__name__] == snapshot


def test_get_geocoding_data_error(caplog, snapshot, respx_mock, settings):
    settings.API_BAN_BASE_URL = "https://geo.foo"
    respx_mock.get(f"{settings.API_BAN_BASE_URL}/search/").respond(200, json=BAN_GEOCODING_API_NO_RESULT_MOCK)

    with pytest.raises(GeocodingDataError):
        geocoding.get_geocoding_data("")
    assert [record for record in caplog.record_tuples if record[0] == geocoding.__name__] == snapshot


@pytest.mark.parametrize("post_code", ["97000", "98999"])
def test_get_geocoding_data_try_without_post_code_if_no_results_for_drom_and_com(
    caplog, snapshot, respx_mock, settings, post_code
):
    settings.API_BAN_BASE_URL = "https://geo.foo"
    respx_mock.get(f"{settings.API_BAN_BASE_URL}/search/").respond(200, json=BAN_GEOCODING_API_NO_RESULT_MOCK)

    with pytest.raises(GeocodingDataError):
        geocoding.get_geocoding_data("HOWELL CENTER", post_code=post_code)
    assert [record for record in caplog.record_tuples if record[0] == geocoding.__name__] == snapshot
