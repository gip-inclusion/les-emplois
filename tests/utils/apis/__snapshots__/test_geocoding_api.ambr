# serializer version: 1
# name: test_get_geocoding_data
  list([
  ])
# ---
# name: test_get_geocoding_data_error
  list([
    tuple(
      'itou.utils.apis.geocoding',
      20,
      'Geocoding error, no result found for `https://geo.foo/search/?q=&limit=1`',
    ),
  ])
# ---
# name: test_get_geocoding_data_try_without_post_code_if_no_results_for_drom_and_com[97000]
  list([
    tuple(
      'itou.utils.apis.geocoding',
      20,
      'Geocoding error, no result found for `https://geo.foo/search/?q=HOWELL+CENTER&limit=1&postcode=97000`',
    ),
    tuple(
      'itou.utils.apis.geocoding',
      20,
      'Geocoding error, no result found for `https://geo.foo/search/?q=HOWELL+CENTER&limit=1`',
    ),
  ])
# ---
# name: test_get_geocoding_data_try_without_post_code_if_no_results_for_drom_and_com[98999]
  list([
    tuple(
      'itou.utils.apis.geocoding',
      20,
      'Geocoding error, no result found for `https://geo.foo/search/?q=HOWELL+CENTER&limit=1&postcode=98999`',
    ),
    tuple(
      'itou.utils.apis.geocoding',
      20,
      'Geocoding error, no result found for `https://geo.foo/search/?q=HOWELL+CENTER&limit=1`',
    ),
  ])
# ---
